### PR TITLE
BUG: fix usage of na_sentinel with sort=True in factorize()

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/pandas-dev/pandas/issues/25409

### CODE REVIEW

File: pandas/core/algorithms.py
Comment (jorisvandenbossche) [inline pandas/core/algorithms.py]: Adding this if/else check to keep the usage of `take_1d` when possible. It was added some time ago to improve performance compared to `safe_sort`, but thus does not handle non-default na_sentinels. 
If we want to keep the code simpler, I can also remove the `take_1d` alltogether, but that would give a performance degradation compared to the last release.

Comment (TomAugspurger) [reply pandas/core/algorithms.py]: This seems reasonable. Alternatively, `safe_sort` could do this try / except?

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: You mean the if / else? (as it would also have to deal with the non-default na_sentinels)

Comment (TomAugspurger) [reply pandas/core/algorithms.py]: Hmm yeah, it would need that as well. But I think it's probably fine here, unless we see this pattern coming up often when using safe_sort.

File: pandas/core/algorithms.py
Comment (jreback) [inline pandas/core/algorithms.py]: woa? this is way repetitive. why is what you have in the try not enough here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/algorithms.py
Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: `take_1d` cannot handle custom na_sentinels. See the discussion in the issue.

Comment (jreback) [reply pandas/core/algorithms.py]: I read the issue. this is fix is way too hacky. A custom sentinel is not a blocking issue for 0.24.2

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: Well, as I said above, I can also simply remove the if/else and try/except completely, by not using `take_1d` if that is preferred. A small performance gain is not worth introducing (now knowingly) a regression in behaviour.

But personally, I also have no problem with just merging this PR as is.

Comment (jreback) [reply pandas/core/algorithms.py]: this is just adding technical debt and am -1 in merging as is

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: Can you then give an indication with what you would be OK, instead of just saying -1? 
Eg I already mentioned that I can remove the usage of take_1d all together (so how it was before 0.23)

Comment (jreback) [reply pandas/core/algorithms.py]: > Eg I already mentioned that I can remove the usage of take_1d all together (so how it was before 0.23)

this sounds fine,  though don't remember why this was changed.

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: Some reasoning is explained here: https://github.com/pandas-dev/pandas/pull/19938#discussion_r173567099 (mainly performance I think)

Comment (jreback) [reply pandas/core/algorithms.py]: ok, then let's just defer this change to 0.25.0 and do this w/o the multiple repetitive try/excepts.

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: If you don't like the extra if/else, but are OK with reverting to only `safe_sort` bu only for 0.25.0: would you be OK with the current PR for 0.24.2, if I directly do a follow-up PR to remove the if/else try/except on master / for 0.25.0 ?

Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: @jreback OK?

Comment (TomAugspurger) [reply pandas/core/algorithms.py]: This would be good for 0.24.2.

What's the duplicate code / tech debt here here? The call to `safe_sort`?

**Comment by jreback** [comment]
not sure why this is that important for 0.24.2. fixing this properly would block the release which doesn't make much sense.

**Comment by TomAugspurger** [comment]
I think it's essentially ready though, right?

**Comment by jreback** [comment]
this is just adding technical debt and am -1 on merging as is. Also the idea of merging a patch then immediately 'fixing' it in master is better, but still, why the urgency. This is a very small edge case, yet adding a horrible hack.

**Comment by jorisvandenbossche** [comment]
> Also the idea of merging a patch then immediately 'fixing' it in master is better, but still, why the urgency.

If you say it is better, do you also mean you are OK with that?

For sure it is not the most important bug fix, but it is a very clear error (the output is simply total nonsense), for which we have a fix that we can just merge, so why not?

**Comment by TomAugspurger** [comment]
Can you explain the tech-debt comment? I still don't quite understand that.

Would you prefer the `try / except` be moved inside its own function? I prefer not to make functions unless they're used in more than one place, since it makes the code harder to follow.

**Comment by jreback** [comment]
> If you say it is better, do you also mean you are OK with that?

no


> Can you explain the tech-debt comment? I still don't quite understand that.>

> Would you prefer the try / except be moved inside its own function? I prefer not to make functions unless they're used in more than one place, since it makes the code harder to follow.

well this is adding multiple try/excepts with the same code. This is horrible.

**Comment by jreback** [comment]
again explain the urgency. this is NOT in any way important at all.

**Comment by jorisvandenbossche** [comment]
As I said above, it is indeed not urgent. But it is a ready fix for a known regression. So again, why not?

You say you want a proper solution. Can you then explain what you mean with this? What do you propose?

> well this is adding multiple try/excepts with the same code. This is horrible.

It is not adding multiple try/excepts. There was one, and there is still one. I only added one if/else case to check for a keyword argument depending on which the try/except should never be tried at all.

**Comment by TomAugspurger** [comment]
How should we proceed then?

**Comment by jreback** [comment]
i see no reason to merge this

this is making already complicated code
way more complicated

**Comment by jorisvandenbossche** [comment]
Our current code is simply wrong. If you don't want the proposed solution here, can you propose an alternative solution?

**Comment by TomAugspurger** [comment]
Joris, maybe try moving the if condition into safe_sort itself? It’ll be the same code, but maybe a new diff would be refreshing.

**Comment by TomAugspurger** [comment]
Actually, doing the `if na_sentiel == -1` check in `safe_sort` isn't quite straightforward, as that accepts list-like and we need array-like.

I'm still not sure where the "way repetitive" code is. The only repetition I see is

```
                uniques, labels = safe_sort(uniques, labels,
                                             na_sentinel=na_sentinel,
                                             assume_unique=True)
```

Am I missing something?

File: pandas/core/algorithms.py
Comment (jreback) [inline pandas/core/algorithms.py]: I don't like the way this is written because it repeats the call to safe_sort. if you can avoid that (easy enough, just use a pass in the except), *then* call safe_sort if needed (assign to uniques, lables as NOne, None initially), then this would be better

**Comment by jorisvandenbossche** [comment]
An additional problem is that `safe_sort` handles more cases, eg it also checks for out of bound labels (which we know we don't have here). So if using `take_1d` inside `safe_sort`, we would need extra code there to also handle out of bound labels when using `take_1d`

File: pandas/core/algorithms.py
Comment (TomAugspurger) [reply pandas/core/algorithms.py]: @jorisvandenbossche do you find this clearer? Personally, I find these dummy variables just for flow control a bit confusing (you have to verify that things are eventually set somewhere). With the way things are written, it's clear that each of the three branches has a `uniques` and `labels`, and it's clear how things are achieved.

If we're really concerned about repeating the call to safe_sort, you can make a closer with the values

```
get_uniques_labels = lambda: safe_sort(uniques, labels, na_sentinel, assume_unique=True)`
```

and call the lambda, but again, that's just unnecessary indirection IMO.

In this case I think just do whatever.

**Comment by jorisvandenbossche** [comment]
I looked a bit further in it. It is possible to move the logic into `safe_sort`, with:

- adding a `check_outofbounds` keyword to disable extra checks (otherwise the performance benefit of `take_1d` is lost)
- fixing `safe_sort` to work for EAs

how it looks like: https://github.com/jorisvandenbossche/pandas/commit/ba944eb09a784605db9a3d51aa154f1f29aa7a5f

Would that be a better solution?

**Comment by TomAugspurger** [comment]
A bit less clear, but perhaps better since all EAs can benefit from it?

File: pandas/core/algorithms.py
Comment (jorisvandenbossche) [reply pandas/core/algorithms.py]: I agree with Tom that this is not necessarily clearer as what it is now.

**Comment by jorisvandenbossche** [comment]
The discussion is mainly about a code style issue (a way to prevent calling ``safe_sort`` twice as Jeff proposes, vs calling the function twice as in the current diff), and not about the actual proposal (falling back to `safe_sort` for the impacted case). And code style is subjective, given that we disagree over which of two is better. But since they are both basically equivalent, I would say this is a minor issue.

With my release manager hat on (but I know, I am not really independent here), I am going to decide not to block this PR over such a minor issue. So I will merge this PR. But, to address the discussion here, I will also open a new PR with my other proposal mentioned above (moving the logic to `safe_sort` and fixing that to handle EAs). But since this touches more parts, I prefer to keep this for 0.25.0.

**Comment by jorisvandenbossche** [comment]
PR is here: https://github.com/pandas-dev/pandas/pull/25696

**Comment by jreback** [comment]
I am -1 on doing this again.  merging bad code is not great. here. There is no urgency on this at all. We hold up PRs all the time on code style.

**Pull request merged**
